core/modules/replica/IngestServerConnection.cc

```diff
@@ -481,26 +486,63 @@ void IngestServerConnection::_loadDataIntoTable() {
                     " VALUES IN (" + to_string(_transactionId) + "))"
             );
-            statements.push_back(
+            dataLoadStatement =
                 "LOAD DATA INFILE " + h.conn->sqlValue(_fileName) +
                     " INTO TABLE " + sqlTable +
-                    " FIELDS TERMINATED BY " + h.conn->sqlValue(string() + _columnSeparator)
-            );
+                    " FIELDS TERMINATED BY " + h.conn->sqlValue(string() + _columnSeparator);
         }
-        for (auto&& statement: statements) {
+        for (auto&& statement: tableMgtStatements) {
             LOGS(_log, LOG_LVL_DEBUG, context << __func__ << "  statement: " << statement);
         }
-        h.conn->execute([&statements](decltype(h.conn) const& conn_) {
-            for (auto&& statement: statements) {
-                conn_->begin();
-                conn_->execute(statement);
-                conn_->commit();
+        LOGS(_log, LOG_LVL_DEBUG, context << __func__ << "  statement: " << dataLoadStatement);
+
+        // Allow retries for the table management statements in case of deadlocks.
+        // Deadlocks may happen when two or many threads are attempting to create
+        // or modify partitioned tables, or at a presence of other threads loading
+        // data into these tables.
+        //
+        // TODO: the experimental limit for the maximum number of retries may need
+        //       to be made unlimited, or be limited by some configurable timeout.
+        int const maxRetries = 1;
+        int numRetries = 0;
+        while (true) {
+            try {
+                h.conn->execute([&tableMgtStatements](decltype(h.conn) const& conn_) {
+                    conn_->begin();
+                    for (auto&& statement: tableMgtStatements) {
+                        conn_->execute(statement);
+                    }
+                    conn_->commit();
+                });
+                break;
+            } catch (database::mysql::LockDeadlock const& ex) {
+                if (h.conn->inTransaction()) h.conn->rollback();
+                if (numRetries < maxRetries) {
+                    LOGS(_log, LOG_LVL_WARN, context << __func__ << "  exception: " << ex.what());
+                    ++numRetries;
+                } else {
+                    LOGS(_log, LOG_LVL_ERROR, context << __func__ << "  maximum number of retries "
+                            << maxRetries << " for avoiding table management deadlocks has been reached."
+                            << " Aborting the file loading operation.");
+                    throw;
+                }
             }
+        }
+
+        // Load table contribution
+        if (dataLoadStatement.empty()) {
+            throw std::runtime_error(context + string(__func__) + "  no data loading statement generated");
+        }
+        h.conn->execute([&dataLoadStatement](decltype(h.conn) const& conn_) {
+            conn_->begin();
+            conn_->execute(dataLoadStatement);
```

Comment:
Could this deadlock with the DDL ALTER TABLE statements? OR is it always DDL that get that error?

Reply:
I haven't see this in the data loading statements. It's always the ALTER TABLE statement which is giving me troubles.